### PR TITLE
TinyASM backend for ASMPatchPC

### PR DESCRIPTION
**Pull request by florianwechsung**
This adds a simple alternative backend to PETSc PCASM for the ASMPatchPC. It builds on #1773 

While PCASM is great for domain decomposition problems where you may want to use other sophisticated solvers on each domain, we are often interested in the limit of many, small patches. In that case, the fastest thing to do is often to just compute the explicit inverse and then apply it each iteration. (Instead of for example computing the LU factorisation and then doing two triangular solves each time). In addition, this avoids the overhead of having KSP and PC objects for every single patch.

For the tests in `test_star_pc.py` this results in a speedup between 2x and 4x.
```
diff --git a/tests/regression/test_star_pc.py b/tests/regression/test_star_pc.py
index ae6f0377..1c4f54dc 100644
--- a/tests/regression/test_star_pc.py
+++ b/tests/regression/test_star_pc.py
@@ -158,13 +158,18 @@ def test_star_equivalence(problem_type):

     nvproblem = NonlinearVariationalProblem(a, u, bcs=bcs)
     star_solver = NonlinearVariationalSolver(nvproblem, solver_parameters=star_params, nullspace=nsp)
+    import time
+    t0 = time.time()
     star_solver.solve()
+    t1 = time.time()
+    print('Time using petsc', t1-t0)
     star_its = star_solver.snes.getLinearSolveIterations()

     try:
         from tinyasm import _tinyasm as tasm  # noqa
         have_tinyasm = True
     except ImportError:
+        print('does not have tinyasm')
         have_tinyasm = False

     if have_tinyasm:
@@ -172,7 +177,10 @@ def test_star_equivalence(problem_type):
         u.assign(0)
         nvproblem = NonlinearVariationalProblem(a, u, bcs=bcs)
         tinyasm_solver = NonlinearVariationalSolver(nvproblem, solver_parameters=star_params, nullspace=nsp)
+        t0 = time.time()
         tinyasm_solver.solve()
+        t1 = time.time()
+        print('Time using TinyASM', t1-t0)
         tinyasm_its = tinyasm_solver.snes.getLinearSolveIterations()
     else:
         tinyasm_its = star_its
```
```
Time using petsc 0.6164448261260986
Time using TinyASM 0.13721990585327148
Time using petsc 2.2604024410247803
Time using TinyASM 0.9665257930755615
Time using petsc 0.5605008602142334
Time using TinyASM 0.13840794563293457
```

The TinyASM code currently lives in https://github.com/florianwechsung/tinyasm, but I'd be happy to move it over to the firedrake organisation. I added an option `--tinyasm` to the install script as well.

### CODE REVIEW

**Comment by florianwechsung** [comment]
This currently fails on jenkins with

```
[2020-07-16T17:48:13.545Z]     CMake Error at CMakeLists.txt:5 (project):
[2020-07-16T17:48:13.545Z]       The CMAKE_C_COMPILER:
[2020-07-16T17:48:13.545Z] 
[2020-07-16T17:48:13.545Z]         mpicc
[2020-07-16T17:48:13.545Z] 
[2020-07-16T17:48:13.545Z]       is not a full path and was not found in the PATH.
[2020-07-16T17:48:13.545Z] 
[2020-07-16T17:48:13.545Z]       Tell CMake where to find the compiler by setting either the environment
[2020-07-16T17:48:13.545Z]       variable "CC" or the CMake cache entry CMAKE_C_COMPILER to the full path to
[2020-07-16T17:48:13.545Z]       the compiler, or to the compiler name if it is in the PATH.

```

Does anyone have any advice? I thought there should be a `mpicc` and `mpicxx` in `firedrake/bin`

**Comment by pefarrell** [comment]
Looking at the standard `Jenkinsfile`, it doesn't pass the `--mpicc` flags (and friends) to `firedrake-install`. Maybe if you add the MPICH packages to `docker/Dockerfile.env` and the flags to the installer script, it might work?

File: scripts/firedrake-install [hunk 1]
```diff
@@ -922,6 +926,32 @@ def build_and_install_libsupermesh():
...
+    else:
+        git_clone("git+https://github.com/florianwechsung/TinyASM.git")
+        tinyasm_changed = True
+
+    if tinyasm_changed:
+        with directory("TinyASM"):
+            # Clean source directory
+            check_call(["git", "reset", "--hard"])
+            check_call(["git", "clean", "-f", "-x", "-d"])
+            run_pip_install(["-e", "."])
```
Comment (wence-) [inline scripts/firedrake-install:950]: We set `MPICC` and similar in the environment that this is run in, but perhaps not with absolute paths?

Comment (wence-) [reply scripts/firedrake-install:950]: The latter probably matters because I think we don't push the `PATH` on?

**Comment by wence-** [comment]
> Looking at the standard Jenkinsfile, it doesn't pass the --mpicc flags (and friends) to firedrake-install. Maybe if you add the MPICH packages to docker/Dockerfile.env and the flags to the installer script, it might work?

That would probably work, but please don't do that, because that's not the way we recommend users on desktop install.

**Comment by wence-** [comment]
In tinyasm's `cmakelists.txt` I see:

```cmake
set(CMAKE_CXX_COMPILER "mpicxx")
set(CMAKE_C_COMPILER "mpicc")
```

Can you default to the `MPICXX` and `MPICC` environment variables instead?

File: tests/regression/test_star_pc.py [hunk 2]
```diff
@@ -162,7 +162,7 @@ def test_star_equivalence(problem_type):
     star_its = star_solver.snes.getLinearSolveIterations()
 
     try:
-        from tinyasm import _tinyasm as tasm
+        from tinyasm import _tinyasm as tasm  # noqa
```
Comment (wence-) [inline tests/regression/test_star_pc.py:165]: You can be more specific in the warning you squash I think. Otherwise this is fine.

**Comment by florianwechsung** [comment]
I updated `CMakeLists.txt` and `pr-merge` is building and passing fine on jenkins. I guess the last thing remaining is the memory leak discussed here https://github.com/florianwechsung/TinyASM/issues/1

File: tests/regression/test_linesmoother.py [hunk 3]
```diff
@@ -38,7 +38,17 @@ def expected(mesh_type):
...
+def backend(request):
+    return request.param
+
+
+def test_linesmoother(mesh, S1family, expected, backend):
+    if backend == "tinyasm":
+        try:
+            from tinyasm import _tinyasm as tasm  # noqa: F401
+        except ImportError:
+            return
```
Comment (wence-) [inline tests/regression/test_linesmoother.py:51]: A cleaner solution:

```
try:
    import tinyasm
    marks = None
except ImportError:
    marks = pytest.mark.skip(reason="No tinyasm")

@pytest.fixture(params=["petscasm", pytest.param("tinyasm", marks=marks)])
 def backend(request):
     return request.param
```

File: firedrake/preconditioners/asm.py [hunk 4]
```diff
@@ -66,6 +89,117 @@ def applyTranspose(self, pc, x, y):
...
+                    if dof <= 0:
+                        continue
+                    off = section.getOffset(p)
+                    # Local indices within W
+                    W_indices = numpy.arange(off*W.value_size, W.value_size * (off + dof), dtype='int32')
+                    indices.extend(V_local_ises_indices[i][W_indices])
+            iset = PETSc.IS().createGeneral(indices, comm=COMM_SELF)
+            ises.append(iset)
+
+        return ises
```
Comment (wence-) [inline firedrake/preconditioners/asm.py:200]: Does it make sense to replicate all of `ASMPatchPC`?

Comment (florianwechsung) [reply firedrake/preconditioners/asm.py:191]: This is building on top of Scott's PR, so the modification to ASMPatchPC is actually much smaller than this diff suggests. However, there was some duplication in the `if backend == 'petscasm`: ... else ... ` part, which I have now reduced.

Comment (ScottMacLachlan) [reply firedrake/preconditioners/asm.py:191]: I think there's some fundamental duplication that we haven't really resolved in my `ASMStarPC` PR (and that's mostly my fault for not getting back to it with any serious effort this week...), trying to see a way forward that doesn't negatively impact `ASMLinesmoothPC`.  I'll take a shot at that now.

File: tests/regression/test_linesmoother.py [hunk 5]
```diff
@@ -38,18 +38,19 @@ def expected(mesh_type):
         return [9, 20]
 
 
-@pytest.fixture(params=["petscasm", "tinyasm"])
+try:
+    import tinyasm  # noqa: F401
+    marks = None
+except ImportError:
+    marks = pytest.mark.skip(reason="No tinyasm")
```
Comment (wence-) [inline tests/regression/test_linesmoother.py:45]: Sorry, I screwed up, this needs to be a set: `marks = {pytest.mark.skip(...)}`

**Comment by wence-** [comment]
I guess this is still waiting on #1773 to go in first. I also suppose the branches that fixed memory leaks in tinyasm need to be merged?

**Comment by florianwechsung** [comment]
Everything on the TinyASM side has now been merged

**Review by wence- — APPROVED** [review]

**Pull request merged**